feat(timepicker): Added min/max attributes to timepicker.

mikeknoll opened the pull request:
Adds attributes for timepicker's min and max time.

wesleycho commented:
That is what `min-date` and `max-date` are for.

wesleycho commented:
I apologize, this is for a different component than I thought - I will review this.

wesleycho commented on src/timepicker/timepicker.js:
```diff
@@ -254,9 +312,14 @@ angular.module('ui.bootstrap.timepicker', [])
+  function addMinutes( date,  minutes ) {
+    var dt = new Date( date.getTime() + minutes * 60000 );
+    var newDate = new Date( date.getTime() ).setHours( dt.getHours(), dt.getMinutes() );
+    return new Date( newDate );
```
Is there any reason why `newDate` can't just get returned? It should already be an independent reference.

mikeknoll replied:
Date.setHours() returns an int, not a Date for some reason.  I returned new Date( newDate ) to convert the int back into a Date.

wesleycho replied:
Oh, I see now.

It would likely be more efficient to do
```
var newDate = new Date( date.getTime() + minutes * 60000 );
newDate.setHours( dt.getHours(), dt.getMinutes() );
return newDate;
```
The `dt` date existed as a temporary structure to figure out how to modify the existing `selected` date, but now we can just return the date of interest

wesleycho commented on src/timepicker/timepicker.js:
```diff
@@ -55,6 +55,50 @@ angular.module('ui.bootstrap.timepicker', [])
+
+  $scope.noIncrementHours = function() {
+    var incrementedSelected = addMinutes( selected, hourStep * 60 );
+    return incrementedSelected > max || (min && incrementedSelected < selected);
```
If !$attrs.max or !$attrs.min, this does not seem to be the behavior we want.

mikeknoll replied:
I'm not sure I understand.  I was following the pattern of the other attributes, such as $attrs.hourStep and $attrs.minuteStep.  If the attribute is present, add a watch for the expression in it.

wesleycho replied:
These functions will run regardless of whether they exist

wesleycho commented:
Thanks for this, this is mostly solid work - I do have two comments that should be addressed before this is merged.

mikeknoll commented:
I believe I addressed your concerns.  I was able to remove an instance of Date in addMinutes(), slightly different than your suggest because it needs to "rollover" from 11:59PM to 12:00AM on the same day, not the next.

I removed the unnecessary conditionals on $attrs.min/max.

wesleycho commented on src/timepicker/timepicker.js:
```diff
@@ -55,6 +55,46 @@ angular.module('ui.bootstrap.timepicker', [])
+
+  $scope.noIncrementHours = function() {
+    var incrementedSelected = addMinutes( selected, hourStep * 60 );
+    return incrementedSelected > max || (min && incrementedSelected < selected);
```
So what I meant was more `return (max && incrementedSelected > max) || incrementedSelected < selected;`, and so on.

mikeknoll replied:
The logic is a bit different than that.  `incrementedSelected > max` is catching the common case, increasing the time would exceed the max.  If max is undefined, `incrementedSelected > max` will be false, which is what we want(to allow the increase).

The other part, `(min && incrementedSelected < selected)` was intended to disable the increment/decrement if the time would rollover(at midnight) if the opposite direction's wouldn't be allowed.  For example, if it is 11:30pm, you cannot increase to 12:30am if the minimum is 1:00am.

This code isn't quite there though.  This just prevents a rollover if the minimum/maximum is defined. For example, if it is 11:30pm, you cannot increase to 12:30am if the minimum is 12:01am.  This should be allowed, as the updated time is still later than the minimum.

I believe the correct code will be `return incrementedSelected > max || (incrementedSelected < selected && incrementedSelected < min);`.  I've manually tested it.  I'll write up some unit tests and submit it.

wesleycho commented on src/timepicker/timepicker.js:
```diff
@@ -55,6 +55,46 @@ angular.module('ui.bootstrap.timepicker', [])
+  };
+
+  $scope.noToggleMeridian = function() {
+    if(selected.getHours() < 13) {
```
Add space here, i.e. `if (`

wesleycho commented on src/timepicker/timepicker.js:
```diff
@@ -177,7 +217,11 @@ angular.module('ui.bootstrap.timepicker', [])
       if ( angular.isDefined(hours) ) {
         selected.setHours( hours );
-        refresh( 'h' );
+        if( selected < min || selected > max) {
```
Change to `if (selected...`

wesleycho commented on src/timepicker/timepicker.js:
```diff
@@ -196,7 +240,11 @@ angular.module('ui.bootstrap.timepicker', [])
       if ( angular.isDefined(minutes) ) {
         selected.setMinutes( minutes );
-        refresh( 'm' );
+        if( selected < min || selected > max) {
```
Change to `if (selected...`

wesleycho commented on src/timepicker/timepicker.js:
```diff
@@ -222,7 +270,13 @@ angular.module('ui.bootstrap.timepicker', [])
         selected = date;
       }
-      makeValid();
+      if( selected < min || selected > max ) {
```
Ditto

wesleycho commented on src/timepicker/timepicker.js:
```diff
@@ -264,19 +324,29 @@ angular.module('ui.bootstrap.timepicker', [])
   };
   $scope.decrementHours = function() {
-    addMinutes( - hourStep * 60 );
+    if(!$scope.noDecrementHours()) {
```
Change to `if (`

wesleycho commented on src/timepicker/timepicker.js:
```diff
@@ -264,19 +324,29 @@ angular.module('ui.bootstrap.timepicker', [])
   };
   $scope.incrementMinutes = function() {
-    addMinutes( minuteStep );
+    if(!$scope.noIncrementMinutes()) {
```
Ditto

wesleycho commented on src/timepicker/timepicker.js:
```diff
@@ -264,19 +324,29 @@ angular.module('ui.bootstrap.timepicker', [])
   };
   $scope.decrementMinutes = function() {
-    addMinutes( - minuteStep );
+    if(!$scope.noDecrementMinutes()) {
```
Ditto

wesleycho commented on src/timepicker/timepicker.js:
```diff
@@ -264,19 +324,29 @@ angular.module('ui.bootstrap.timepicker', [])
   };
   $scope.toggleMeridian = function() {
-    addMinutes( 12 * 60 * (( selected.getHours() < 12 ) ? 1 : -1) );
+    if(!$scope.noToggleMeridian()) {
```
Ditto

wesleycho commented:
Once these changes are in, this is good to go.

mikeknoll commented:
I updated the style, and changed the noIncremenentHours() etc... to disable the time step in a direction if when time rolls over, the result exceeds the limit for movement in the opposite direction.

wesleycho commented:
LGTM - merging shortly with documentation & code style fixes.
